src/hotspot/share/interpreter/oopMapCache.cpp

```diff
@@ -582,11 +588,21 @@ void OopMapCache::enqueue_for_cleanup(OopMapCacheEntry* entry) {
   }
 }
 
-// This is called after GC threads are done and nothing is accessing the old_entries
-// list, so no synchronization needed.
+bool OopMapCache::has_cleanup_work() {
+  return Atomic::load(&_old_entries) != nullptr;
+}
```

Comment:
Right. I have not realized the service thread timed wait addition was recent. So it would become a problem if we pull this patch to the JDK release where service thread needs to be explicitly notified. We need to do this carefully, though, since we don't want to acquire `ServiceLock` all that often or inundate the thread with requests. So I added a simple notification when queue is populated with the first element.